### DISCUSSION TITLE
Is there a way to upload a gcode file and start the print?

### DISCUSSION THREAD

**Discussion by trohnjavolta** [greghesp/ha-bambulab#307] [Q&A]
I love this intergration and would really like to use it, however with current firmware it seems only one mqtt connection is allowed which means I cannot use both this integration and orca slicer to send and start a print.
I found this: https://www.wolfwithsword.com/bambulab-remote-print-from-home-assistant/

But this describes a whole other approach, using nodered and mqtt, right?
Is there any other way to upload gcode and start a print?

Currently I keep swapping sd card around which gets annoying.

**Comment by AdrianGarside** [comment]
Not currently. And the node red based integration is still talking to the printer via mqtt so it'll hit the same problem. Current lan mode is broken in the P1 printers. Not using lan mode and configuring this integration to use bambu cloud connection is the only guaranteed functional way to allow multiple connections.

**Comment by trohnjavolta** [comment]
It would make the slicer connection obsolete, so there'd be only one connection right?
Or do both, hacs integration and this remote printing flow need seperate connections?

**Comment by AdrianGarside** [comment]
Are you envisioning that you slice the file and then save the gcode from the slicer for separate upload by the integration?

**Reply by trohnjavolta** [reply]
Exactly this, yes! Do you think there's a way to do so?

**Reply by AdrianGarside** [reply]
Yes it’s possible. You can use ftps to upload to the printer and there is a command to send to tell it to print. The tricky part is if an AMS is involved as the command needs to tell it how to map the filament in the sliced file to the filaments in the AMS. I’m not sure how to craft a UX for that in home assistant or if you even could.

**Reply by trohnjavolta** [reply]
I don't have an AMS...yet, so yeah, some integration would be awesome. But I understand that this is a fault from bambulabs, it wouldn't make any sense to impelemnt a lan only mode but restrict it to one connection only at the same time.

**Reply by AdrianGarside** [reply]
I am noodling on whether I could add a 'polling' mode where the integration doesn't maintain a persistent connection but instead just connects, asks for the current state, disconnects on a timer - or perhaps simply user initiated (via the existing 'force refresh' button). This _might_ be a cheap change to make and allow some use of the integration in a way that works with the P1 firmware bug. The fidelity of the experience won't be as good though but perhaps good enough? And way cheaper than creating a good gcode upload experience would be even assuming it went well...

**Reply by AdrianGarside** [reply]
I wonder if there's a way to know if something is being viewed in a dashboard and auto-trigger the poll for fresh data only when the dashboard is being actively viewed?

**Comment by jneilliii** [comment]
I've been able to do this with the Simply Print integration I did for bambu. They use ftps for the connection to upload.

**Reply by trohnjavolta** [reply]
Can you elaborate on this? Do you mean this? https://simplyprint.io/compatibility/bambu-lab-p1s
I guess I could run it in docker on my homeserver: https://github.com/SimplyPrint/SimplyPi-Docker

Is this a Octoprint fork? Still there is no Orca Slicer integraton or is there? Would be cool to press one button in Orca Slicer and have the gcode transfered via ftps and then start the print, all without bambu cloud.

**Reply by jneilliii** [reply]
Sorry for the late reply. It's not an OctoPrint fork, but they are utilizing a separate integration to their platform. I've begun my OctoPrint integration for Bambu printers and have gotten it uploading to my X1C, but I think the SimplyPrint team has had to modify the implementation to work with P1P/S and A1 printers due to different ftps implementation is what I understand. You can check out my OctoPrint integration [here](https://github.com/jneilliii/OctoPrint-BambuPrinter), you upload to the printer as if it was a connected SD card. I'm reporting temperatures and have print control at the moment, albeit with many quirks.
